### DISCUSSION TITLE
Implemented ACL with fastapi

### DISCUSSION THREAD

**Discussion by ManiMozaffar** [fastapi/fastapi#10157] [Show and tell]
I have implemented ACL with fastapi, which I think can be quite useful for permission, so you can use same endpoint with same query but with different user roles, or dynamically add role to your application with just a json.

https://github.com/ManiMozaffar/fast-student

This part my favorite!
```python
def get_permission_setting():
    ACL_MAPPER: dict[RoutesEnum, dict[UserRoles, PermissionGrants]] = {
        RoutesEnum.ADD_STUDENT: {
            UserRoles.ADVISOR: PermissionGrants.NOT_ALLOWED,
            UserRoles.STUDENT: PermissionGrants.NOT_ALLOWED,
            UserRoles.TEACHER: PermissionGrants.RELATED_CLASSROOM,
        },
        RoutesEnum.DELETE_STUDENT: {
            UserRoles.ADVISOR: PermissionGrants.NOT_ALLOWED,
            UserRoles.STUDENT: PermissionGrants.NOT_ALLOWED,
            UserRoles.TEACHER: PermissionGrants.RELATED_CLASSROOM,
        },
        RoutesEnum.EXPEL_STUDENT: {
            UserRoles.ADVISOR: PermissionGrants.NOT_ALLOWED,
            UserRoles.STUDENT: PermissionGrants.NOT_ALLOWED,
            UserRoles.TEACHER: PermissionGrants.RELATED_CLASSROOM,
        },
        RoutesEnum.UPDATE_STUDENT: {
            UserRoles.ADVISOR: PermissionGrants.NOT_ALLOWED,
            UserRoles.STUDENT: PermissionGrants.NOT_ALLOWED,
            UserRoles.TEACHER: PermissionGrants.RELATED_CLASSROOM,
        },
        RoutesEnum.READ_STUDENT: {
            UserRoles.ADVISOR: PermissionGrants.IS_ALLOWED,
            UserRoles.STUDENT: PermissionGrants.RELATED_CLASSROOM,
            UserRoles.TEACHER: PermissionGrants.RELATED_CLASSROOM,
        },
    }
    for router in ACL_MAPPER:
        ACL_MAPPER[router][UserRoles.ADMIN] = PermissionGrants.IS_ALLOWED
    return ACL_MAPPER
```

If you liked this idea, maybe you can upvote:
https://github.com/tiangolo/fastapi/discussions/10156

**Comment by tarekasishm** [comment]
I usually prefer to use a battle-tested solution for AuthN. Take a look at [FastAPI Casbin Integration](https://github.com/pycasbin/fastapi-authz). It lets you apply all the authorization techniques you may need.

**Reply by ManiMozaffar** [reply]
Hmm I have quite good reason to not use middleware and do what I did in this case:

1. Using middleware, you make middleware to rely on the "path" of request, which may change, and it violates law of Demeter.
2. You are forced to apply ACL on all routes, on public routers maybe you prefer not to, we don't have "ACL" on something that is entirely public.

It'd be awesome if framework would allow me to DI the router object itself, to kinda have that awareness when I'm adding it to a router.
